packages/astro/src/runtime/server/render/component.ts

```diff
@@ -349,67 +382,154 @@ function sanitizeElementName(tag: string) {
 	return tag.trim().split(unsafe)[0].trim();
 }
 
-async function renderFragmentComponent(result: SSRResult, slots: ComponentSlots = {}) {
+async function renderFragmentComponent(
+	result: SSRResult,
+	slots: ComponentSlots = {}
+): Promise<RenderInstance> {
 	const children = await renderSlotToString(result, slots?.default);
-	if (children == null) {
-		return children;
-	}
-	return markHTMLString(children);
+	return {
+		render(destination) {
+			if (children == null) return;
+			destination.write(children);
+		},
+	};
 }
 
 async function renderHTMLComponent(
 	result: SSRResult,
 	Component: unknown,
 	_props: Record<string | number, any>,
 	slots: any = {}
-) {
+): Promise<RenderInstance> {
 	const { slotInstructions, children } = await renderSlots(result, slots);
 	const html = (Component as any)({ slots: children });
 	const hydrationHtml = slotInstructions
-		? slotInstructions.map((instr) => stringifyChunk(result, instr)).join('')
+		? slotInstructions.map((instr) => chunkToString(result, instr)).join('')
 		: '';
-	return markHTMLString(hydrationHtml + html);
+	return {
+		render(destination) {
+			destination.write(markHTMLString(hydrationHtml + html));
+		},
+	};
 }
 
-export function renderComponent(
+async function renderAstroComponent(
+	result: SSRResult,
+	displayName: string,
+	Component: AstroComponentFactory,
+	props: Record<string | number, any>,
+	slots: any = {}
+): Promise<RenderInstance> {
+	const instance = await createAstroComponentInstance(result, displayName, Component, props, slots);
+
+	// Eagerly render the component so they are rendered in parallel
+	const chunks: RenderDestinationChunk[] = [];
+	const temporaryDestination: RenderDestination = {
+		write: (chunk) => chunks.push(chunk),
+	};
+	await instance.render(temporaryDestination);
+
+	return {
+		render(destination) {
+			// The real render function will simply pass on the results from the temporary destination
+			for (const chunk of chunks) {
+				destination.write(chunk);
+			}
+		},
+	};
+}
+
+export async function renderComponent(
 	result: SSRResult,
 	displayName: string,
 	Component: unknown,
 	props: Record<string | number, any>,
 	slots: any = {}
-): Promise<ComponentIterable> | ComponentIterable | AstroComponentInstance {
+): Promise<RenderInstance> {
```

Comment:
`renderComponent` is the core API for rendering any component. Now it will return a promise so it's rendered in parallel, and return a `RenderInstance` - a new concept added, but basically an object with `{ render: (destination) => void }`. (destination is the `RenderDestination`)